code/parachain/frame/staking-rewards/src/lib.rs

```diff
@@ -774,145 +838,252 @@ pub mod pallet {
 				*inflation = inflation.safe_add(additional_inflation)?;
 			}
 
+			let fnft_asset_account =
+				T::FinancialNft::asset_account(&fnft_collection_id, &fnft_instance_id);
+
 			// TODO (vim): transfer the staked amount to the NFT account and lock it
 			// TODO (vim): Transfer the shares with share asset ID to the Financial NFT account and
 			// lock it.
 			T::Assets::transfer(
 				rewards_pool.asset_id,
 				who,
-				&Self::pool_account_id(&stake.reward_pool_id),
+				&fnft_asset_account,
 				amount,
 				keep_alive,
 			)?;
+			T::Assets::mint_into(rewards_pool.share_asset_id, &fnft_asset_account, amount)?;
 			RewardPools::<T>::insert(stake.reward_pool_id, rewards_pool);
-			Stakes::<T>::insert(position, stake);
-			Self::deposit_event(Event::<T>::StakeAmountExtended { position_id: position, amount });
-			Ok(position)
+			Stakes::<T>::insert(fnft_collection_id, fnft_instance_id, stake);
+			Self::deposit_event(Event::<T>::StakeAmountExtended {
+				amount,
+				fnft_collection_id,
+				fnft_instance_id,
+			});
+			Ok((fnft_collection_id, fnft_instance_id))
 		}
 
 		#[transactional]
-		fn unstake(who: &Self::AccountId, position_id: &Self::PositionId) -> DispatchResult {
+		fn unstake(
+			who: &Self::AccountId,
+			(fnft_collection_id, fnft_instance_id): &Self::PositionId,
+		) -> DispatchResult {
 			let keep_alive = false;
-			let mut stake =
-				Stakes::<T>::try_get(position_id).map_err(|_| Error::<T>::StakeNotFound)?;
-			ensure!(who == &stake.owner, Error::<T>::OnlyStakeOwnerCanUnstake);
+
+			let mut stake = Stakes::<T>::try_get(fnft_collection_id, fnft_instance_id)
+				.map_err(|_| Error::<T>::StakeNotFound)?;
+
+			let owner = T::FinancialNft::owner(fnft_collection_id, fnft_instance_id)
+				.ok_or(Error::<T>::FnftNotFound)?;
+
+			ensure!(who == &owner, Error::<T>::OnlyStakeOwnerCanUnstake);
```

Comment:
I guess this check belongs in the extrinsic rather than here, for all operations by the fNFT owners

Reply:
Not sure I follow here - do you mean move the owner check to the extrinsic body?

Reply:
Yeah, the rational being that these validations prevent the trait being used from the other pallets unless they fake it

Reply:
Would there ever be a case where these operations would be done by someone who isn't the owner?

Reply:
So, should we generally strive to keep all identity validation on the `pallet::call` level? @vimukthi-git

Reply:
@benluelo Not at the level of extrinsics that can be called by any account.

Reply:
@PoisonPhang Yes. trait impl must be usable across all pallets as a library when needed without hacky work arounds.

Reply:
btw this does not need to be handled in this PR. I raised this just to discuss.